### PR TITLE
ems: Tidy ems.sys loading

### PR DESCRIPTION
**Pull request by andrewbird**
1/ Add check for ems.sys already loaded
2/ Remove unused message strings from ems.S
3/ Remove cr,lf and eom constants from ems.S, they are unnecessary
4/ Remove set of header Unit field (only valid on a block device driver)
5/ Set the proper exit status for device driver load failure

### CODE REVIEW

File: src/dosext/misc/emm.c [hunk 1]
```diff
@@ -273,33 +273,44 @@ static void set_map_registers(const struct emm_reg *buf, int pages);
...
 
+static int ems_init_done = 0;
+
 void
 ems_helper(void) {
   switch (LWORD(ebx)) {
   case 0:
     E_printf("EMS Init called!\n");
     if (!config.ems_size) {
+      LWORD(eax) = 1;	/* report error */
```
Comment (stsp) [inline src/dosext/misc/emm.c]: This is the old-style error reporting.
How about using CARRY instead?

File: src/dosext/misc/emm.c [hunk 2]
```diff
@@ -273,33 +273,44 @@ static void set_map_registers(const struct emm_reg *buf, int pages);
...
       clear_IF();
       return;
     }
+
     if (HI(ax) < DOSEMU_EMS_DRIVER_VERSION) {
       warn("EMS driver too old, consider updating %i->%i\n",
         HI(ax), DOSEMU_EMS_DRIVER_VERSION);
       com_printf("EMS driver too old, consider updating.\n");
     }
+    ems_init_done = 1;	/* make sure the driver isn't loaded twice */
```
Comment (stsp) [inline src/dosext/misc/emm.c]: I think it would be better to do
the EMS check in ems.sys itself.
If EMS is present, it should not
load. Its because if you use the
native himem.sys and try to load
it twice, it gracefully fails without
any help from dosemu.

File: src/dosext/misc/emm.c (hunk 1 above)
Comment (andrewbird) [inline src/dosext/misc/emm.c]: It's to fit with ems.S doing this at line 235
```ASM
orb %al, %al
jz 1f
```
Before this patch it's only set on success, which means on error it has whatever value was left by the call. But I'm happy to change, i just worry about going off and changing loads of unrelated stuff unnecessarily.

File: src/dosext/misc/emm.c (hunk 2 above)
Comment (andrewbird) [inline src/dosext/misc/emm.c]: I thought about that, but there is the case where EMS is disabled and XMS is available the driver gets loaded and XMS handlers do get setup. So I suppose I'd need to do the same for XMS too?

Comment (andrewbird) [inline src/dosext/misc/emm.c]: Mmm, I'm not sure about that any more. With 
~~~
$_ems = (off)
~~~
I see message in DOS about EMS not being available, a message about dosemu XMS being installed and yet the driver and handler is not installed as the 'andb NoEMS, %al' is true and the jump to Error is done. It's as if the 'call HookHimem' is returning AL == 1, which I think means XMS not installed. I don't know enough about EMS/XMS DOS memory to say what the correct behaviour should be when we turn off EMS in dosemu.conf. Can you check the HookHimem call does the right thing in that case please?

Comment (andrewbird) [inline src/dosext/misc/emm.c]: >I think it would be better to do
the EMS check in ems.sys itself.
If EMS is present, it should not
load. Its because if you use the
native himem.sys and try to load
it twice, it gracefully fails without
any help from dosemu.

But what would you use to check in ems.sys, there's no persistent memory between attempted loads of the device driver, unless you want to test the interrupt vector or the sniff the device driver chain for name 'EMMXXXX0'?

Assuming we find another EMS driver already installed somehow, does that mean the XMS part of this driver shouldn't install either?

Comment (stsp) [inline src/dosext/misc/emm.c]: > the 'call HookHimem' is returning AL == 1, which I think means XMS not installed.

As the comment says,
`/* internal driver installed - go out with error */`
IIRC this means that internal driver does
not need any assistance from ems.sys, so
it gets unloaded, but the `XMS_HELPER_XMS_INIT`
is being called, which is enough. In case
of an external himem.sys, ems.sys needs
to stay, so in that case it won't return error.

> But what would you use to check in ems.sys

EMS spec is quite clear on what is to
check, namely this:
http://www.delorie.com/djgpp/doc/rbinter/ix/67/40.html

> Assuming we find another EMS driver already installed
> somehow, does that mean the XMS part of this driver shouldn't install either?

I am pretty sure yes.
Because then you don't even need to
see if you found this one or another one -
it doesn't even matter. If something is
pretending to be ems driver, we go out.

Comment (stsp) [inline src/dosext/misc/emm.c]: > EMS spec is quite clear on what is to
> check, namely this:

No, sorry, the example in the spec does this:
```
          char emm_installed()

          {
              char *EMM_device_name = "EMMXXXX0";
              char *int_67_device_name_ptr;

              /*--------------------------------------------------------*/
              /* AH = DOS get interrupt vector function.                */
              /*--------------------------------------------------------*/
              input_regs.h.ah = 0x35;
              
              /*--------------------------------------------------------*/
              /* AL = EMM interrupt vector number.                      */
              /*--------------------------------------------------------*/
              input_regs.h.al = EMM_INT;
              intdosx (&input_regs, &output_regs, &segment_regs);
              
              /*--------------------------------------------------------*/
              /* Upon return ES:0Ah points to location where            */
              /* device name should be.                                 */
              /*--------------------------------------------------------*/
              int_67_device_name_ptr = build_ptr (segment_regs.es, 0x0A);
              
              /*--------------------------------------------------------*/
              /* Compare memory with EMM device name.                   */
              /*--------------------------------------------------------*/
              if (memcmp (EMM_device_name, int_67_device_name_ptr,
                                             DEVICE_NAME_LENGTH) == 0)
                  return (TRUE);
              else
                  return (FALSE);
          }
```

**Comment by andrewbird** [comment]
Thanks for the pointer to the EMMXXXX0 test. I've implemented it now, and it can detect a prior load of our ems.sys. I want to load another DOS EMS provider, so I tried EMM386.EXE (PC-DOS 7) but it complains about already being in protected mode. Is there something else that will provide the EMS driver I can test with?

**Comment by stsp** [comment]
Not really.
Maybe we can support XT-style
EMS with the native driver, but they
are extinct these days.

**Comment by stsp** [comment]
> are extinct these days.

But google reveals some emm286.
If you have enough time, you can get
it to work. :)

**Comment by andrewbird** [comment]
Cool, using emm286.exe from [pccorner.com](https://www.pcorner.com/list/UTILITY) with himem.sys loaded beforehand, ems.sys detects the previous EMM and declines to install.

~~~
Starting PC DOS...
 
 
EMM286 - EMS 4.0 LIMulator
Version 1.2 (c) 1993 Jjex Software
WARNING: An EMS manager has already been loaded.
~~~

**Comment by stsp** [comment]
Wow!
But does emm286 work?
We don't even need ems.sys then? :)

**Comment by stsp** [comment]
Hmm, it does indeed work!
But it does not enable UMBs with an
external himem.sys. So we still need
our ems.sys. In that case it should
stay resident but without registering
the device and without hooking on
int67.

**Comment by stsp** [comment]
Thanks!
Now lets get emm286 supported! :)

**Comment by andrewbird** [comment]
> So we still need
our ems.sys. In that case it should
stay resident but without registering
the device and without hooking on
int67.

I think maybe the way to stay resident without providing EMS, is to not register the EMMXXXX0 name but something else. I suspect that's probably what other memory managers do.  I found [this](http://collaboration.cmc.ec.gc.ca/science/rpn/biblio/ddj/Website/articles/DDJ/1994/9409/9409m/9409m.htm) quite interesting. It has the following text:
~~~
A memory manager with EMS disabled (such as EMM386 with "NOEMS" or QEMM with "EMS=NONE")
does not install "EMMXXXX0" because EMS is not active. Instead, a device with a slightly
different name ("EMMQXXX0" for EMM386 and QEMM, "QMMXXXX0" for 386MAX) is installed.
~~~
I wonder if when we establish that we don't want to install EMS, we can change the name in the header?

**Comment by andrewbird** [comment]
This also makes interesting reading. [gemmis](http://dgi_il.tripod.com/gemmis.txt). It seems to explain why the ems.sys advertises itself as supporting IOCTL functions. Does Windows work in 386 mode with Dosemu2? I don't pretend to understand much about Windows, but perhaps we need to fill out some data in the IOCTL read, rather than just calling the Dummy function?

**Comment by stsp** [comment]
I added the initial emm286 support. :)

> I wonder if when we establish that we don't want to install EMS,
> we can change the name in the header?

It seems DOS takes the device name directly
from the file, so altering it in init code leads to
strange things...
I think the aforementioned EMM managers
traverse the DOS device list and alter the name
there. Can you give it a try? :)

> Does Windows work in 386 mode with Dosemu2?

By skipping the ring-0 kernel completely.
So gemmis is not needed for it.

**Comment by stsp** [comment]
But emm286 takes quite a lot of DOS
memory and I don't think we can relocate
it to UMB without loading it _after_ our
ems.sys, in which case it won't load...

**Comment by stsp** [comment]
It is possible to disable EMS in dosemu.conf,
load ems.sys (with altered device name) before
emm286, and then load emm286 into UMB with
`devicehigh`.
So we need `NOEMS` to our ems.sys. :)
But I can't test windows with it because our
msdos.c calls directly into dosemu's EMM.

**Comment by andrewbird** [comment]
> It seems DOS takes the device name directly
from the file, so altering it in init code leads to
strange things...

Do you mean it reads the header and stores the name, before calling the Intr function 0 (init)? After Init returns success it then uses the stored name when adding to the device driver chain?


> I think the aforementioned EMM managers
traverse the DOS device list and alter the name
there.

The trouble with that is that whilst the Init function is still running our device driver won't have been added to the device chain.

There may be another way of doing it. The device driver header has a pointer to the next driver, which is usually set to -1. I wonder if we can add two device drivers in a single 'device=' from config.sys. More complex for sure, two Init functions, and so two returned results. If so there's a potential to load the HiMEM hook in one, the EMS in the other.

I'll test altering the name in the Init function here, just to see if I can recreate the strange effects, as if it works it would be the most simple method.

**Comment by stsp** [comment]
> Do you mean it reads the header and stores the name, before calling the Intr function 0 (init)?

It seems so, at least for the small testing I did.

> I'll test altering the name in the Init function here, just to see if I can recreate the strange effects

The full test is to load emm286 before ems.sys,
then ems.sys alters its device name. Then you
run mft from qemm and in DOS -> Drivers you
see that the name is unchanged on ems.sys but
is changed on emm286...

I know what to do. :) We just need to split the
ubm functionality into an umb.sys. We can put
the common code into some `umb.inc` so that
also ems.sys can include it for compatibility.

**Comment by andrewbird** [comment]
~~~
 
Conventional memory detail:
 
  Segment                Size        Name         Type
  -------          ----------------  -----------  --------
   00000              1,024    (1K)               Interrupt Vector
   00040                256    (0K)               ROM Communication Area
   00050                512    (1K)               DOS Communication Area
   00070              2,800    (3K)  IBMBIO       System Data
                                        CON       System Device Driver
                                        AUX       System Device Driver
                                        PRN       System Device Driver
                                        CLOCK$    System Device Driver
                                        A: - E:   System Device Driver
                                        COM1      System Device Driver
                                        LPT1      System Device Driver
                                        LPT2      System Device Driver
                                        LPT3      System Device Driver
                                        COM2      System Device Driver
                                        COM3      System Device Driver
                                        COM4      System Device Driver
   0011F             42,096   (41K)  IBMDOS       System Data
   00B66             12,848   (13K)  IBMBIO       System Data
                        160    (0K)     AMMXXXX0  Installed Driver EMS
~~~

This is what mem shows on PC-DOS 7. The header has it defined as 'EMMXXXX0', just before Init() return true I change its first character to 'A'. That said I just read your 'fuil test' and I'm nowhere near as confident!

Do you want to try the two device drivers in one file thing? My 'Writing MS-DOS Device Drivers' book says it's how the five std drivers are implemented in IBMBIO.COM.

**Comment by stsp** [comment]
```
diff --git a/src/env/commands/ems.S b/src/env/commands/ems.S
index 3148076..5d5ed4b 100644
--- a/src/env/commands/ems.S
+++ b/src/env/commands/ems.S
@@ -51,8 +51,8 @@ Header:
                                # (char, supports IOCTL strings (it doesn't!)
        .word   Strat           # ptr to strategy routine
        .word   Intr            # ptr to interrupt service routine
-EMSStr:
-       .ascii  "EMMXXXX0"      # logical-device name
+EMSName:
+       .ascii  "12345678"      # changed on init
 
 # the Strat and Intr routines are entered with a "far call".  I don't
 # know how to tell gas that as I would in Turbo Assembler, so I just
@@ -65,6 +65,9 @@ OldInt2f:     .long 0
 EMSBios:       .long 0
 NoEMS:         .byte 0
 AlreadyLoaded: .byte 0
+EMSStr:
+       .ascii  "EMMXXXX0"      # logical-device name
+AltEMM:                .ascii  "EMMDEMU2"
 
 Dispatch:
        .word   Init            # initialize driver
@@ -267,17 +270,24 @@ Init:
        cld
        repe    cmpsb
 
-       pop     %di
-       pop     %es
-
        jne     22f
        movb    $1, AlreadyLoaded
+       lea     AltEMM, %si
+       lea     EMSName, %di
+       movw    $EMSLEN, %cx
+#      rep     movsb
+
+       pop     %di
+       pop     %es
        movb    $9, %ah
        movw    $EmsAlreadyLoadedMsg, %dx
        int     $0x21
        jmp     1f
 
 22:
+       pop     %di
+       pop     %es
+
 # Check if Dosemu has enabled EMS support
        movb    $DOS_HELPER_EMS_HELPER, %al
        movb    $DOSEMU_EMS_DRIVER_VERSION, %ah
```

Here's my test case.
Please try it with the "full test-case" description above.

**Comment by stsp** [comment]
> Do you want to try the two device drivers in one file thing?

What will this solve?
Maybe DOS will initially take all of them from
the header?
I think having `umb.sys` is cleaner as `ems.sys`
have already grown a lot and became complex.

**Comment by stsp** [comment]
Note: I now edited the patch to uncomment
`rep movsb` that was forgotten.

**Comment by andrewbird** [comment]
>What will this solve?
Maybe DOS will initially take all of them from
the header?

That's fine, but depending upon whether the respective Init() return true or false determines whether it gets installed to the chain. So if dosemu.conf disables ems, the EMMXXXX0 is not loaded into memory, but discarded. But the UMBXYZZ0 driver can still be loaded independently of it.

**Comment by stsp** [comment]
Yes, this can solve the problem, but at
the cost of adding `NOEMS` option support.
Surely you are not going to always disable
ems in dosemu.conf. So I am absolutely
sure having the separate driver is simpler,
but if you think otherwise, feel free to do it
the hard way. :)

**Comment by andrewbird** [comment]
I didn't try the full test, but even on my simple mem one, I can see that you need to set the ES 

```asm
        movb    $1, AlreadyLoaded                                               
                                                                                
+        push %cs                                                                
+       pop  %es                                                                
                                                                                
       lea     AltEMM, %si                                                      
       lea     EMSName, %di                                                     
```

**Comment by stsp** [comment]
Also with your approach `config.sys` may
look clumsy. Say you have something like this:
```
device=c:\dos\himem.sys^M
device=f:\umb.sys^M
devicehigh=c:\emm286.exe 8000^M
devicehigh=f:\ems.sys^M
```

That looks quite fine and allows you to
easily disable `emm286` in an interactive
startup (f8).
With your approach similar config will look like this:
```
device=c:\dos\himem.sys^M
device=f:\ems.sys NOEMS^M
devicehigh=c:\emm286.exe 8000^M
devicehigh=f:\ems.sys^M
```

**Comment by andrewbird** [comment]
Yes, looks horrid. I certainly hadn't considered that you'd want to load ems.sys twice, once with NOEMS, and secondly for another reason.

So in your scenario what does each provide?

device     | service
------------ | -------------
himem.sys | ?
umb.sys | UMB
emm286.exe | EMS
ems.sys | ?

**Comment by stsp** [comment]
OK, thanks, with setting FS this works.
So I pushed the patch in.

**Comment by stsp** [comment]
> So in your scenario what does each provide?

himem.sys provides xms.
ems.sys does not provide anything by default,
but, as I wrote above, I can use interactive
startup (f8) and skip emm286.exe, in which
case ems.sys will start to provide EMS.

**Comment by andrewbird** [comment]
Loading ems.sys twice after himem.sys without any other drivers now gives this
~~~
Starting PC DOS...
 
UMB support enabled.
dosemu EMS driver rev 0.7 installed.
WARNING: An EMS manager has already been loaded.
Unable to hook into himem.sys, UMB disabled.
Make sure himem.sys is loaded right before ems.sys in your config.sys.
~~~

I think we are circumventing our check for previous invocation now. Yes it's getting complicated :smiley:

**Comment by andrewbird** [comment]
With config.sys as foillows and default dosemu.conf

~~~
lastdrive=Z
device=himem.sys
device=dosemu\ems.sys
~~~

~~~
Press any key to continue...
 
 
  EMS Handle    EMS Name      Size
  ----------    --------    -------
        0       SYSTEM         384K
        1                       64K
 
  Total Expanded (EMS)                  8,781,824   (8,576K)
  Free Expanded (EMS)                   8,323,072   (8,128K)
 
 
 
  Total Extended (XMS)                  8,454,144   (8,256K)
  Free Extended (XMS)                   8,388,608   (8,192K)
 
  Memory accessible using INT 15h               0       (0K)
  Largest executable program size         587,664     (574K)
  Largest free upper memory block               0       (0K)
  The high memory area is available.
 
  XMS Version  3.00; Driver Version  3.15
  EMS Version   4.0; Page Frame at E000
~~~

Should I be seeing any UMBs?

**Comment by stsp** [comment]
```
UMB support enabled.
dosemu EMS driver rev 0.7 installed.
```
This is from first invocation.

```
WARNING: An EMS manager has already been loaded.
Unable to hook into himem.sys, UMB disabled.
Make sure himem.sys is loaded right before ems.sys in your config.sys.
```
This is from second one.

What's wrong do you see with this?

**Comment by stsp** [comment]
> Should I be seeing any UMBs?

If you do `DOS=HIGH,UMB` then yes.

**Comment by andrewbird** [comment]
Yes, the messages are fine.

**Comment by andrewbird** [comment]
>DOS=HIGH,UMB

Thanks, that works.

**Comment by stsp** [comment]
> I think we are circumventing our check for previous invocation now. Yes it's getting complicated 😃

Yes, there are now 2 totally independent
device drivers. IMHO it would be better to
have them split.

> Thanks, that works.

Have you never done that before?
So you were using dos without UMBs, congrats. :)

**Comment by andrewbird** [comment]
Probably about 30 years ago! To be honest I don't use DOS except for testing Dosemu. For me it's just fun.

**Comment by andrewbird** [comment]
I had read that the 'Limulators' (286 version of EMM386.EXE) were very slow, so just for completeness I ran this benchmark [movetest](https://www.pcorner.com/list/DIAGS/EMSTST10.ZIP/INFO/). The results are significant such that I don't think anyone would want to run one instead of Dosemu's internal EMS.

Limulator [EMM286.EXE](https://www.pcorner.com/list/UTILITY/EMM286.ZIP/INFO/)
------------
~~~
C:\>movetest
EMS 4.0 Move Block Performance Test 1.02 ■ Douglas Boling
 
This test measures the time an Expanded memory manager takes to move
256 Kbytes of data.
 
Please wait...
 
Using 64 byte records,
 Expanded to Conventional = 15.54 Sec.
 Conventional to Conventional = 09.45 Sec.
 Conventional to Expanded = 12.80 Sec.
 Expanded to Expanded = 19.61 Sec.
 
Using 128 Kbyte records,
 Expanded to Conventional = 00.00 Sec.
 Conventional to Conventional = 00.00 Sec.
 Conventional to Expanded = 00.05 Sec.
 Expanded to Expanded = 00.00 Sec.
~~~

Dosemu2
------------
~~~
C:\>movetest
EMS 4.0 Move Block Performance Test 1.02 ■ Douglas Boling
 
This test measures the time an Expanded memory manager takes to move
256 Kbytes of data.
 
Please wait...
 
Using 64 byte records,
 Expanded to Conventional = 00.10 Sec.
 Conventional to Conventional = 00.10 Sec.
 Conventional to Expanded = 00.10 Sec.
 Expanded to Expanded = 00.10 Sec.
 
Using 128 Kbyte records,
 Expanded to Conventional = 00.00 Sec.
 Conventional to Conventional = 00.00 Sec.
 Conventional to Expanded = 00.00 Sec.
 Expanded to Expanded = 00.00 Sec.
 ~~~

**Comment by stsp** [comment]
Likely not much slower than simple memcpy().
I am not sure what exactly this test does:
EMM uses 16K pages, and I don't know
what is "64 byte records" and "128 Kbyte records".
And on "128 Kbyte records" there is no
difference, anyway.
I'd like to get this thing supported.
Even if no one uses it, that would still
result in an overall improvements
(eg umb.sys split).

**Comment by stsp** [comment]
> I had read that the 'Limulators' (286 version of EMM386.EXE) were very slow

I think emm386.exe is also a limulator,
just a fast one. :) Everything that does
not represent a physical EMS board,
was called a limulator, IIRC.

**Comment by andrewbird** [comment]
The fact that small record size transfers are much slower than large ones to copy the same amount of data suggests an overhead to copy a record. Probably we would see the difference between EMM286.EXE and Dosemu's ems.sys on the large record test if more iterations were done. I have no interest either way, but since I stumbled upon it I figured it might be a nice data point.

**Comment by stsp** [comment]
But I don't even know what does this
"record size" mean in the EMS terms.
Do you have any idea?
I.e. currently I don't even know what
does this test ever tests.

**Comment by andrewbird** [comment]
Nope, I'm afraid not.
